scilpy/reconst/raw_signal.py

```diff
@@ -0,0 +1,117 @@
+# -*- coding: utf-8 -*-
+
+import logging
+
+import numpy as np
+
+from dipy.core.sphere import Sphere
+from dipy.reconst.shm import sf_to_sh
+
+from scilpy.utils.bvec_bval_tools import (check_b0_threshold,
+                                          is_normalized_bvecs, normalize_bvecs,
+                                          identify_shells)
+
+
+def compute_sh_coefficients(dwi, gradient_table, sh_order=8,
+                            basis_type='descoteaux07', smooth=0.006,
+                            use_attenuation=False, force_b0_threshold=False,
+                            mask=None):
+    """Fit a diffusion signal with spherical harmonics coefficients.
+
+    Parameters
+    ----------
+    dwi : nib.Nifti1Image object
+        Diffusion signal as weighted images (4D).
+    gradient_table : GradientTable
+        Dipy object that contains all bvals and bvecs.
+    sh_order : int, optional
+        SH order to fit, by default 8.
+    smooth : float, optional
+        Lambda-regularization coefficient in the SH fit, by default 0.006.
+    basis_type: str
+        Either 'tournier07' or 'descoteaux07'
+    use_attenuation: bool, optional
+        If true, we will use DWI attenuation. [False]
+    force_b0_threshold : bool, optional
+        If set, will continue even if the minimum bvalue is suspiciously high.
+    mask: nib.Nifti1Image object
+        Binary mask. Only data inside the mask will be used for computations
+        and reconstruction.
+
+    Returns
+    -------
+    sh_coeffs : np.ndarray with shape (X, Y, Z, #coeffs)
+        Spherical harmonics coefficients at every voxel. The actual number
+        of coefficients depends on `sh_order`.
+    """
+
+    # Extracting infos
+    b0_mask = gradient_table.b0s_mask
+    bvecs = gradient_table.bvecs
+    bvals = gradient_table.bvals
+
+    # Checks
+    if not is_normalized_bvecs(bvecs):
+        logging.warning("Your b-vectors do not seem normalized...")
+        bvecs = normalize_bvecs(bvecs)
+    check_b0_threshold(force_b0_threshold, bvals.min())
+
+    # Ensure that this is on a single shell.
+    shell_values, _ = identify_shells(bvals)
+    shell_values.sort()
+    if shell_values.shape[0] != 2 or shell_values[0] > 20.:
```

Comment:
`shell_values[0] > 20.` We should use the threshold in check_b0_threshold, or use `bvals.min()` if --force_b0_threshold was given